### PR TITLE
feat(as): add planned task resource support

### PR DESCRIPTION
**Pull request by jinyangyang222**
**What this PR does / why we need it**:
add planned task resource support
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
add planned task resource support
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST="./huaweicloud/services/acceptance/as" TESTARGS="-run TestAccPlannedTask_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccPlannedTask_basic -timeout 360m -parallel 4
=== RUN   TestAccPlannedTask_basic
=== PAUSE TestAccPlannedTask_basic
=== CONT  TestAccPlannedTask_basic
--- PASS: TestAccPlannedTask_basic (146.38s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        146.417s

make testacc TEST="./huaweicloud/services/acceptance/as" TESTARGS="-run TestAccPlannedTask_timedTask"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccPlannedTask_timedTask -timeout 360m -parallel 4
=== RUN   TestAccPlannedTask_timedTask
=== PAUSE TestAccPlannedTask_timedTask
=== CONT  TestAccPlannedTask_timedTask
--- PASS: TestAccPlannedTask_timedTask (128.65s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        128.707s
```

### CODE REVIEW

File: docs/resources/as_planned_task.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+subcategory: "Auto Scaling"
+---
+
+# huaweicloud_as_planned_task
+
+Manages an AS planned task resource within HuaweiCloud.
+
+## Example Usage
+
+### AS scheduled planned Task
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]: `scheduled Task` or `planned Task`?

File: docs/resources/as_planned_task.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+```hcl
+variable "scaling_group_id" {}
+
+resource "huaweicloud_as_planned_task" "test" {
+  scaling_group_id = var.scaling_group_id
+  name             = "test"
+
+  scheduled_policy {
+    launch_time     = "2023-11-30T12:00Z"
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]:
```suggestion
    launch_time = "2023-11-30T12:00Z"
```

File: docs/resources/as_planned_task.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  When not set or is empty, the planned task is scheduled execution.
+  After setting, the planned task is periodic execution. The valid values are as follows:
+  + **DAILY**: by day periodic execution.
+  + **WEEKLY**: by week periodic execution.
+  + **MONTHLY**: by month periodic execution.
+
+* `recurrence_value` - (Optional, String) Specifies the frequency at which planned task are triggered.
+  Required only when `recurrence_type` is **WEEKLY** or **MONTHLY**.
+  + When `recurrence_type` is **WEEKLY**, The valid value range is **1-7**.
+  + When `recurrence_type` is **MONTHLY**, The valid value range is **1-31**.
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]:
```suggestion
  + When `recurrence_type` is **WEEKLY**, The valid value ranges from `1` to `7`.
  + When `recurrence_type` is **MONTHLY**, The valid value ranges from `1` to `31`.
```

File: docs/resources/as_planned_task.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+* `scaling_group_id` - (Required, String, ForceNew) Specifies the AS scaling group ID.
+  Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the name of planned task.
+
+* `scheduled_policy` - (Required, List) Specifies the scheduled or periodic policy for AS planned task.
+  The [scheduled_policy](#AS_ScheduledPolicy) structure is documented below.
+
+* `instance_number` - (Required, List) Specifies the numbers of scaling group instance for AS planned task.
+  At least set one of `max`, `min` or `desire` parameters.
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]: Mark in the specific parameter description.

File: go.sum [hunk 5]
```diff
@@ -24,6 +24,8 @@ github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5/go.mod h1:wHh0iHkY
 github.com/census-instrumentation/opencensus-proto v0.2.1/go.mod h1:f6KPmirojxKA12rnyqOA5BBL4O983OfeGPqjHWSTneU=
 github.com/chnsz/golangsdk v0.0.0-20231103032017-ed0e526911d5 h1:CTjY2iKJqzsA8zrPYttHasMVQZUDZHPXZPCW+fNSv0A=
 github.com/chnsz/golangsdk v0.0.0-20231103032017-ed0e526911d5/go.mod h1:Erm4hDWxXgAdbkG3+hhJFgRzEL1TvvcroWzw2Gax4uI=
+github.com/chnsz/golangsdk v0.0.0-20231110072139-c889f680af60 h1:lFB8MfBdVVFfFS4rVHzBWuX6ihGEsaWoPkj0ckpOdfs=
+github.com/chnsz/golangsdk v0.0.0-20231110072139-c889f680af60/go.mod h1:Erm4hDWxXgAdbkG3+hhJFgRzEL1TvvcroWzw2Gax4uI=
```
Comment (Lance52259) [inline go.sum]: `go mod tidy` please

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+				d.Set("scaling_group_id", task.GroupID),
+				d.Set("scheduled_policy", flattenPlannedTaskSchedulePolicy(task.ScheduledPolicy)),
+				d.Set("instance_number", flattenPlannedTaskInstanceNumber(task.InstanceNumber)),
+				d.Set("created_at", task.CreateTime),
+			)
+			return diag.FromErr(mErr.ErrorOrNil())
+		}
+	}
+
+	return nil
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]:
```suggestion
	return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{}, "AS planned task")
```

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+			return diag.FromErr(err)
+		}
+
+		updateOpts.ScheduledPolicy = &scheduledtasks.ScheduledPolicy{
+			LaunchTime:      scheduledPolicy.LaunchTime,
+			RecurrenceType:  scheduledPolicy.RecurrenceType,
+			RecurrenceValue: scheduledPolicy.RecurrenceValue,
+			StartTime:       scheduledPolicy.StartTime,
+			EndTime:         scheduledPolicy.EndTime,
+		}
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]:
```suggestion
		updateOpts.ScheduledPolicy = &scheduledPolicy
```
`instance_number` ditto

File: docs/resources/as_planned_task.md (hunk 1 above)
Comment (jinyangyang222) [reply docs/resources/as_planned_task.md]: fixed.

File: docs/resources/as_planned_task.md (hunk 2 above)
Comment (jinyangyang222) [reply docs/resources/as_planned_task.md]: fixed.

File: docs/resources/as_planned_task.md (hunk 3 above)
Comment (jinyangyang222) [reply docs/resources/as_planned_task.md]: fixed.

File: docs/resources/as_planned_task.md (hunk 4 above)
Comment (jinyangyang222) [reply docs/resources/as_planned_task.md]: fixed.

File: go.sum (hunk 5 above)
Comment (jinyangyang222) [reply go.sum:26]: fixed.

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go (hunk 6 above)
Comment (jinyangyang222) [reply huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]: fixed.

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go (hunk 7 above)
Comment (jinyangyang222) [reply huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]: fixed.

File: docs/resources/as_planned_task.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+The `instance_number` block supports:
+
+* `max` - (Optional, Int) Specifies the maximum number of instances for the scaling group. The default value is `0`.
+
+* `min` - (Optional, Int) Specifies the minimum number of instances for the scaling group. The default value is `0`.
+
+* `desire` - (Optional, Int) Specifies the expected number of instances for the scaling group.
+  The default value is `0`.
+
+-> At least set one of `max`, `min` or `desire` parameters.
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]: According to your code logic, if `min` is set but `desire` is not set, an error will be reported. Also, if `desire` is set but `max` is not set, an error will be reported.

File: docs/resources/as_planned_task.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID.
+
+* `created_at` - The (UTC) creation time of the event source, in RFC3339 format.
+
+## Import
+
+The AS planned task can be imported using the related `scaling_group_id` and their `id`, separated by a slash, e.g.
```
Comment (Lance52259) [inline docs/resources/as_planned_task.md]:
```suggestion
The AS planned task can be imported using the related `scaling_group_id` and `id`, separated by a slash, e.g.
```

File: huaweicloud/services/acceptance/as/resource_huaweicloud_as_planned_task_test.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    launch_time      = "12:00"
+    recurrence_type  = "MONTHLY"
+    start_time       = "2025-12-30T12:00Z"
+    end_time         = "2026-01-30T12:00Z"
+    recurrence_value = "25"
+  }
+
+  instance_number {
+    max    = "2"
+    min    = "0"
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/as/resource_huaweicloud_as_planned_task_test.go]:
```suggestion
    max = "2"
    min = "0"
```

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"max": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "The maximum number of instances for the scaling group",
+						},
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]:
```suggestion
						"max": {
							Type:        schema.TypeString,
							Optional:    true,
							Description: "The maximum number of instances for the scaling group",
							AtLeastOneOf: []string{"min", "desire"},
						},
```
maybe append an `AtleastOneOf`

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+	"github.com/chnsz/golangsdk"
+	"github.com/chnsz/golangsdk/openstack/autoscaling/v1/scheduledtasks"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+)
+
+var (
+	PlannedTaskRecurrencePeriods = []string{"DAILY", "WEEKLY", "MONTHLY"}
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]:
```suggestion
	plannedTaskRecurrencePeriods = []string{"DAILY", "WEEKLY", "MONTHLY"}
```
only this package used.

File: huaweicloud/services/acceptance/as/resource_huaweicloud_as_planned_task_test.go (hunk 10 above)
Comment (jinyangyang222) [reply huaweicloud/services/acceptance/as/resource_huaweicloud_as_planned_task_test.go]: done.

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go (hunk 11 above)
Comment (jinyangyang222) [reply huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]: fixed.

File: huaweicloud/services/as/resource_huaweicloud_as_planned_task.go (hunk 12 above)
Comment (jinyangyang222) [reply huaweicloud/services/as/resource_huaweicloud_as_planned_task.go]: fixed.

**Pull request merged**